src/cassette/TsxImage.cc

```diff
@@ -0,0 +1,480 @@
+    * [#4B] MSX implementation of KCS (Kansas City Standard) Block
+    * [#5A] Glue Block
+
+    Blocks detected and skipped but not yet supported or not very useful
+    for MSX tapes:
+
+    * [#11] Turbo Speed Block (Turbo speed)
+    * [#14] Pure data Block
+    * [#15] Direct recording Block
+    * [#19] Generalized data block
+    * [#21] Group start Block
+    * [#22] Group end Block
+    * [#23] Jump Block
+    * [#24] Loop start Block
+    * [#25] Loop end Block
+    * [#26] Call sequence Block
+    * [#27] Return sequence Block
+    * [#31] Message Block
+*/
+
+/*
+	Standard MSX Blocks (#4B) can be forced to 3600 bauds, reduction of pilot
+	pulses, and after minimal block pauses.
+	The Turbo blocks (#10) can't be forced beyond 140% over real speed but they
+	are not forced to maintain compatibility with all Turbo loaders.
+*/
+static const bool     ULTRA_SPEED       = true;		//3600 bauds / short pilots / minimal block pauses
+
+// output settings
+static const unsigned Z80_FREQ          = 3500000;	// 3.5 Mhz
+static const unsigned OUTPUT_FREQ       = 58900;	// ~ = Z80_FREQ*4/TSTATES_MSX_PULSE
+static const float    TSTATES_MSX_PULSE = 238.f;
+
+// headers definitions
+static const byte TSX_HEADER   [ 8] = { 'Z','X','T','a','p','e','!', 0x1a};
+static const byte ASCII_HEADER[10]  = { 0xEA,0xEA,0xEA,0xEA,0xEA,0xEA,0xEA,0xEA,0xEA,0xEA };
+static const byte BINARY_HEADER[10] = { 0xD0,0xD0,0xD0,0xD0,0xD0,0xD0,0xD0,0xD0,0xD0,0xD0 };
+static const byte BASIC_HEADER[10]  = { 0xD3,0xD3,0xD3,0xD3,0xD3,0xD3,0xD3,0xD3,0xD3,0xD3 };
+
+#define tstates2bytes(tstates)	( (tstates) * OUTPUT_FREQ / Z80_FREQ )
+
+
+inline uint16_t toLe16(byte *data)
+{
+	return (((uint16_t)data[1])<<0) | (((uint16_t)data[0])<<8);
+}
+inline uint24_t toLe24(byte *data)
+{
+	return (((uint24_t)data[2])<<0) | (((uint24_t)data[1])<<8) | (((uint24_t)data[0])<<16);
+}
+inline uint32_t toLe32(byte *data)
+{
+	return (((uint32_t)data[3])<<0) | (((uint32_t)data[2])<<8) | (((uint32_t)data[1])<<16) | (((uint32_t)data[0])<<24);
+}
+
+
+TsxImage::TsxImage(const Filename& filename, FilePool& filePool, CliComm& cliComm)
+{
+	setFirstFileType(CassetteImage::UNKNOWN);
+	convert(filename, filePool, cliComm);
+}
+
+int16_t TsxImage::getSampleAt(EmuTime::param time)
+{
+	static const Clock<OUTPUT_FREQ> zero(EmuTime::zero);
+	unsigned pos = zero.getTicksTill(time);
+	return pos < output.size() ? output[pos] * 256 : 0;
+}
+
+EmuTime TsxImage::getEndTime() const
+{
+	Clock<OUTPUT_FREQ> clk(EmuTime::zero);
+	clk += unsigned(output.size());
+	return clk.getTime();
+}
+
+unsigned TsxImage::getFrequency() const
+{
+	return OUTPUT_FREQ;
+}
+
+void TsxImage::fillBuffer(unsigned pos, int** bufs, unsigned num) const
+{
+	size_t nbSamples = output.size();
+	if (pos < nbSamples) {
+		for (auto i : xrange(num)) {
+			bufs[0][i] = (pos < nbSamples)
+			           ? output[pos] * 256
+			           : 0;
+			++pos;
+		}
+	} else {
+		bufs[0] = nullptr;
+	}
+}
+
+void TsxImage::writePulse(uint32_t tstates) {
```

Comment:
The rest of the file uses a new line before opening brace. Style is broken here.